core/modules/editor/factory.js

```diff
@@ -257,6 +277,11 @@ function editTextWidgetFactory(toolbarEngine,nonToolbarEngine) {
 		var editInfo = this.getEditInfo();
 		if(text !== editInfo.value) {
 			editInfo.update(text);
+		} else if(text === editInfo.value) {
```

Comment:
The second "else if" could just be "else".

Are we certain that this code will never clear the error state without previously having set it? It seems as though there could be false fires here. We want to avoid that because removing the class means hitting the DOM.

Reply:
`setError` would be called on every update of the value anyway (maybe `updateValidationStatus` would be a good name). If I read `addClass`/`removeClass` correctly this will only lead to a DOM update if the value changes. I think it should be the task of the engine to optimize these calls if this is necessary.

Are DOM reads so bad that it would be worthwhile to have a cache (an additional variable) inside the engine to prevent these? I normally hesitate to do such things because having the same state duplicated in two places has bitten me more than once.

PS: I'll convert the `else if` into an `else`. Sorry for the badly executed last minute change.

Reply:
In general we try hard to minimise the work in the refresh method. Here I think I'm just not getting the logic; the concern is that we appear to selectively engage the error mode but disengage it unconditionally.

Reply:
Hello @Jermolene, please take a look at the new `updateValidationStatus` code. I've split the code into a `setValidationStatus` function within `factory.js` that can be called unconditionally and an `updateValidationStatus` function within the engine that is only called by `setValidationStatus` if the validation status has changed.

The use of the `setValidationStatus` function keeps the code clean, centralizing the change detection and the update of the state variable into one location.

This should minimize the work that is done is `saveChanges` by only calling the (potentially costly) `updateValidationStatus` function if the validation status has changed.

As an additional benefit `setValidationStatus` checks if the engine implements `updateValidationStatus` and skips the call if the function is not implemented. That should make it backward compatible with plugins that supply an engine, that does not implement `updateValidationStatus`.